### ISSUE TITLE
Issue with Aggregate Responses

### ISSUE THREAD

**Issue by kliegeois** [SNLComputation/Albany#638]
There is an issue with Aggregate Responses. If I understand the `AggregateScalarResponseFunction` correctly, its purpose is to have several responses without coupling them by summing them together for instance.

In Thyra, there are capabilities to support multiple response functions.
For instance, it is possible to access the g space of the j-th response: `get_g_space(j)`.

Finally, in `Piro_PerformAnalysis`, there is an option to specify the response index which should be used for the optimization problem using `Response Vector Index` in the .yaml.

Based on this context, I assumed that the Aggregate Responses are intended to work as follows:
- Aggregate `n` responses together,
- Allow the possibility to perform the analysis on the j-th response alone.

However, the Aggregate Responses do not work like that currently.

This is due to lines 98-99 of src/responses/Albany_ResponseFactory.cpp:
```
    if(name == "Aggregate Responses")
      responses.push_back(rcp(new Albany::AggregateScalarResponseFunction(comm, scalar_responses)));
```
which `push_back` all the Aggregate Responses in `responses[0]`.

This implies that:
```
app->getNumResponses()
```
returns 1.

As a consequence, using a .yaml as the one of FO_GIS_betaT
```
    Response Functions:
      Number: 2
      Response 0: Surface Velocity Mismatch
      ResponseParams 0: 
        Regularization Coefficient: 1.00000000000000000e+00
        Asinh Scaling: 1.00000000000000000e+01
        Scaling Coefficient: 1.0
        Cubature Degree: 3
        Field Name: Velocity
      Response 1: Surface Mass Balance Mismatch
      ResponseParams 1: 
        Regularization Coefficient: 1.00000000000000000e+00
        H Coefficient: 1.00000000000000000e+00
        SMB Coefficient: 1.00000000000000000e+02
        Scaling Coefficient: 1.0
        Cubature Degree: 3
        Field Name: Velocity

```
which results in Aggregate Responses (default value if `Collection Method` is not specified) and we if use:
```
        Response Vector Index: 1
```
this results in the error:
```
ROL::ThyraProductME_Objective_SimOpt::value

p=0: *** Caught standard std::exception of type 'Teuchos::Exceptions::InvalidParameter' :

 /home/kim/dev/Albany_all/Albany/src/Albany_ModelEvaluator.cpp:431:
 
 Throw number = 1
 
 Throw test that evaluated to true: l >= app->getNumResponses() || l < 0
 
 
 Error!  Albany::ModelEvaluator::get_g_space():  Invalid response index l = 1
 
*** Teuchos::StackedTimer::report() - Remainder for a level will be ***
*** incorrect if a timer in the level does not exist on every rank  ***
*** of the MPI Communicator.                                        ***
```

Another occurrence of this error is when we build Trilinos with TEUCHOS_DEBUG enabled and when we use `Response Vector Index: 0`.
When this is the case, Thyra performs tests to check the compatibility of the vector spaces.

During this compatibility check, it gets the vector space of the first response: `get_g_space(0)`.

In Albany, this will return the vector space of `responses[0]` which is an Aggregate Response, and, in the case of the .yaml listed above, it will return a vector space of dimension 2.
This vector space is incompatible with other vector spaces used during the Piro_PerformAnalysis computation:

```
ROL performing gradient check 1 of 1, at parameter initial guess
Checking Accuracy of Objective Gradient 
ROL::ThyraProductME_Objective_SimOpt::update, Either The State Or The Parameters Changed
ROL::ThyraProductME_Objective_SimOpt::update, Signaling That Parameter Changed
ROL::ThyraProductME_Objective_SimOpt::value
 basal_friction has changed!
ROL::ThyraProductME_Objective_SimOpt::gradient_1

p=0: *** Caught standard std::exception of type 'Thyra::Exceptions::IncompatibleVectorSpaces' :

 /home/kim/local/trilinos_all/Albany/include/Thyra_ModelEvaluatorHelpers.hpp:518:
 
 Throw number = 1
 
 Throw test that evaluated to true: !l_isCompatible
 
 Thyra::DefaultModelEvaluatorWithSolveFactory{thyraModel='Albany::ModelEvaluator',W_factory='Thyra::AztecOOLinearOpWithSolveFactory{precFactory=Thyra::Ifpack2PreconditionerFactory}'}
 
 Error, the following vector spaces are not compatible:
 
  DgDx(0)^T.domain() : Thyra::TpetraVectorSpace<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >{globalDim=1,localSubDim=1,localOffset=0,comm=Teuchos::MpiComm<long>{size=1,rank=0,rawMpiComm=0x55e9e2f31380}}
 
  g_space(0) : Thyra::TpetraVectorSpace<double, int, long long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >{globalDim=2,localSubDim=2,localOffset=0,comm=Teuchos::MpiComm<long>{size=1,rank=0,rawMpiComm=0x55e9e2f31380}}
 
*** Teuchos::StackedTimer::report() - Remainder for a level will be ***
*** incorrect if a timer in the level does not exist on every rank  ***
*** of the MPI Communicator.                                        ***
```

I think that, when we use Aggregate Response, the lines 98-99 of src/responses/Albany_ResponseFactory.cpp:

```
    if(name == "Aggregate Responses")
      responses.push_back(rcp(new Albany::AggregateScalarResponseFunction(comm, scalar_responses)));
```
should be replaced by something like this:

```
    if(name == "Aggregate Responses")
      for (int i=0; i<aggregated_responses.size(); i++)
        responses.push_back(scalar_responses[i]);
```

Could you confirm that this is the expected behavior of the Aggregate Response?

**Comment by bartgol** [comment]
@kliegeois I think you raise a good question: what's the point of disguising N responses as a single response?

**Comment by kliegeois** [comment]
@bartgol You summarized the issue in one sentence!

**Comment by mperego** [comment]
This is something that has been there before I started using Albany. My understanding is that it was used for grouping responses the same way you group scalar parameters.. so instead of having separate responses you have a vector of responses.. but it might not be needed any longer.

**Comment by bartgol** [comment]
If Albany stored a single response, then this would make sense. But since albany stores a vector of responses, then I guess the grouping can already be done in the vector.

I'm ok with purging this class. Of course, adding proper code in the response factory that creates a vector of responses.

**Comment by ikalash** [comment]
> This is something that has been there before I started using Albany. My understanding is that it was used for grouping responses the same way you group scalar parameters.. so instead of having separate responses you have a vector of responses.. but it might not be needed any longer.

I echo this statement.  I am not sure what was the reason for this design.  I am the one who converted the response processing in Main_Solve.cpp to Thyra while ago.  I vaguely remember some issue with > 1 responses, perhaps b/c there was something missing in ME with > 1 response?  I think this is the comment: https://github.com/SNLComputation/Albany/blob/master/src/Main_Solve.cpp#L282-L288 .  It may or may not be relevant.  Anyway, I am OK redesigning things to avoid this issue.

**Comment by kliegeois** [comment]
I am currently trying to push back each response individually.
I will let you know how it goes.

**Comment by mperego** [comment]
> 
> 
> I am currently trying to push back each response individually.
> I will let you know how it goes.

@kliegeois , I've been looking into that. Please wait. The issue is that we need to change all the Regression in all input files to do that. I'll add a more detailed message later.

**Comment by kliegeois** [comment]
> > I am currently trying to push back each response individually.
> > I will let you know how it goes.
> 
> @kliegeois , I've been looking into that. Please wait. The issue is that we need to change all the Regression sections in all input files to do that. I'll add a more detailed message later.

@mperego, thanks for looking into that! That is exactly what I was currently modifying; I was trying something such as:

```
  Regression Results 0: 
    Number of Comparisons: 1
    Test Values: [1.54847000000000001e+00]
    Relative Tolerance: 1.00000000000000005e-04
    Number of Sensitivity Comparisons: 3
    Sensitivity Test Values 0: [1.56217999999999990e+00]
    Sensitivity Test Values 1: [1.66396000000000015e+01]
    Sensitivity Test Values 2: [2.01329000000000002e+00]
  Regression Results 1: 
    Number of Comparisons: 1
    Test Values: [1.69218499999999992e+01]
    Relative Tolerance: 1.00000000000000005e-04
    Number of Sensitivity Comparisons: 3
    Sensitivity Test Values 0: [1.28730000000000011e+00]
    Sensitivity Test Values 1: [1.14771699999999992e+01]
    Sensitivity Test Values 2: [1.01805999999999996e+00]
  Regression Results 2: 
    Number of Comparisons: 1
    Test Values: [2.20519499999999979e+00]
    Relative Tolerance: 1.00000000000000005e-04
    Number of Sensitivity Comparisons: 3
    Sensitivity Test Values 0: [8.52306000000000008e-01]
    Sensitivity Test Values 1: [8.75256599999999985e+00]
    Sensitivity Test Values 2: [9.61594999999999978e-01]
```
I will wait.

**Comment by mperego** [comment]
Right, exactly. Sorry I was not expecting you to look into it, otherwise I could have save you (or myself, hehe) some time. I think we should discuss this within the Albany team. I'm fine changing the input file, but while we are doing that I will advocate for a cleaning of the way responses, parameters and regression results are handled. There are typically two choices for each of them, often labeled as the new and old way.. b.t.w the "new" way was there more than 7 years ago, I think. It's not sustainable and confusing.
My idea is to have only one supported format and make it simpler when possible.
As an example, for the regression of the responses, if we assume the responses are always scalar, we can simplify a bit the regression section (no need to use array or to report the `Number of Comparisons` for Responses).

**Comment by kliegeois** [comment]
Sorry, I was not expecting you to look into it this weekend either. Otherwise, I would have shared my findings.

Yes, I think that this issue is a good opportunity if we want to revisit the format supported.

I will stop working on this issue until a new format has been decided.

**Comment by bartgol** [comment]
I agree that it's time to fully commit to the "new" way of prescribing parameters/responses. Legacy code is ok while people transition to the new standard, but then it ought to be purged.

**Comment by kliegeois** [comment]
I am currently working on it.
I will create a PR in which I will only modify some of the .yaml files to discuss the modifications.
When everything is fine, the other tests will be updated.

**Comment by kliegeois** [comment]
The PR #640 has just been created.
Please, let me know if you have any concerns.

**Comment by kliegeois** [comment]
This issue can be closed as #640  has been merged.
